pkg/clustermesh/common/remote_cluster.go

```diff
@@ -132,6 +135,19 @@ func (rc *remoteCluster) releaseOldConnection() {
 }
 
 func (rc *remoteCluster) restartRemoteConnection() {
+	connectionEstablished := false
+	// Applicable to kvstoremesh only:
+	// We execute a goroutine to implement a timeout on connection.
+	// If the timeout has been reached and a connection hasn't been established, we
+	// remove the cluster from the SyncState to prevent it from blocking readiness forever.
+	go func() {
+		<-time.After(option.Config.KVstoreConnectivityTimeout)
+		rc.mutex.Lock()
+		defer rc.mutex.Unlock()
+		if !connectionEstablished {
```

Comment:
Probably a bit more straightforward would be to have `connectionEstablished` as channel and just wait on both timeout and receiving information on channel about connectionEstablished - we will also exit early this goroutine if connections will be established + no need for locking

Reply:
> But then I realized that this will only happen if we fail to write the cluster config to the kvstore, in which case we probably do want to block readiness.

There's also the case where we fail to retrieve the cluster config, which should be probably treated as failing to connect to etcd in terms of not blocking readiness (it typically happens when there's something wrong in the cluster name configuration).

Reply:
What do you think about using a channel for `connectionEstablished` and closing is at the end of the DoFunc before successful exit? There's a chance we might hit the timeout and un-block a cluster that's actually going to synchronize (i.e. writing the cluster config to the kvstore and we're waiting for the ready channel to be closed), but this seems like an edge-case and a pretty benign one - and will make the logic much simpler to follow

Reply:
How about https://github.com/cilium/cilium/pull/30361#issuecomment-1912286336? It should be even simpler as it doesn't require to modify anything in the common package.